zkevm-circuits/src/evm_circuit/execution/return.rs

```diff
@@ -24,25 +46,313 @@ impl<F: Field> ExecutionGadget<F> for ReturnGadget<F> {
         let opcode = cb.query_cell();
         cb.opcode_lookup(opcode.expr(), 1.expr());
 
-        // TODO: Other constraints are ignored now for RETURN to serve as a
-        // mocking terminator
+        let offset = cb.query_cell();
+        let length = cb.query_rlc();
+        cb.stack_pop(offset.expr());
+        cb.stack_pop(length.expr());
+        let range = MemoryAddressGadget::construct(cb, offset, length);
+
+        let is_success = cb.call_context(None, CallContextFieldTag::IsSuccess);
+        cb.require_equal(
+            "if is_success, opcode is RETURN. if not, opcode is REVERT",
+            opcode.expr(),
+            is_success.expr() * OpcodeId::RETURN.expr()
+                + not::expr(is_success.expr()) * OpcodeId::REVERT.expr(),
+        );
+
+        // There are 4 cases non-mutually exclusive, A to D, to handle, depending on if
+        // the call is, or is not, a create, root, or successful. See the specs at
+        // https://github.com/privacy-scaling-explorations/zkevm-specs/blob/master/specs/opcode/F3RETURN.md
+        // for more details.
+        let is_create = cb.curr.state.is_create.expr();
+        let is_root = cb.curr.state.is_root.expr();
+
+        // These are globally defined because they are used across multiple cases.
+        let copy_rw_increase = cb.query_cell();
+        let copy_rw_increase_is_zero = IsZeroGadget::construct(cb, copy_rw_increase.expr());
+
+        // Case A in the specs.
+        cb.condition(is_create.clone() * is_success.expr(), |cb| {
+            cb.require_equal(
+                "increase rw counter once for each memory to bytecode byte copied",
+                copy_rw_increase.expr(),
+                range.length(),
+            );
+        });
+        cb.condition(
+            is_create.clone() * is_success.expr() * not::expr(copy_rw_increase_is_zero.expr()),
+            |_cb| {
+                // TODO: copy_table_lookup for contract creation.
+            },
+        );
 
-        Self { opcode }
+        // Case B in the specs.
+        cb.condition(is_root.expr(), |cb| {
+            cb.require_next_state(ExecutionState::EndTx);
+            cb.call_context_lookup(
+                false.expr(),
+                None,
+                CallContextFieldTag::IsPersistent,
+                is_success.expr(),
+            );
```

Comment:
I added the `StepStateTransition` here. 

I think that we should never be using `StepStateTransition::any()` /  `Transition::Any`?

Reply:
> I added the `StepStateTransition` here.
> 
> I think that we should never be using `StepStateTransition::any()` / `Transition::Any`?

When using `any` we just don't enforce any constraint for the next step on that field, meaning that the prover is free to set it to any value.  When `any` is used, less constraints are defined, which can be good? (To be sincere, I'm not really sure how adding or removing these constraints affects the performance, maybe it's negligible?)

The argument in favor of using `any` is that in some execution states, we constraint which execution state the next one will be; and that next step may already constrain some of its own fields.  After a root return, we can only encounter `EndTx`. I believe `EndTx` only uses `log_id`, `gas_left` and `rw_counter`.  After `EndTx` we can encounter `BeginTx` or `EndBlock`.
`EndBlock` only uses `rw_counter` and `call_id`.
`BeginTx` is going to constraint the following step fields to fixed values (not deltas): https://github.com/privacy-scaling-explorations/zkevm-circuits/blob/2235254416aefd1bc9b5592e4e06cd58f8516e2a/zkevm-circuits/src/evm_circuit/execution/begin_tx.rs#L200

So in conclusion, I think here we can leave some fields to `any`, as they will either be unused or constrained to a fixed value later.  But constraining them all is also correct.  And in general, it seems a good idea to avoid leaving unconstrained witnesses as much as possible, because it gives code with less chances of having bugs.  For example in the snippet I copied I realize I missed constraining `log_id` to be `Same`, which would be a bug.  But I'm unable to reason about possible performance costs.

Reply:
In general, I agree that we shouldn't add redundant constraints. However, I think it's better for clarity to constrain every field of the `StepStateTransition`. This way, we don't need to reason about what fields of it EndTx, EndBlock, or BeingTx use.

Reply:
Sounds good to me!